In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("Inventory_SupplyChain.csv")

In [3]:
df.shape

(10069, 17)

In [4]:
df.head()

,Date,Region,Product ID,Product Name,Category,Supplier,Warehouse,Order Status,Units Sold,Inventory Level,Transportation Cost,Order Accuracy,Lead Time (Days),Backorder,Average Inventory,Warehouse Capacity,COGS
0,2025-06-11,West,P06754,Tablet,electronics,Supplier A,Warehouse 2,Fulfilled,124.0,1993.0,16045.78,True,1.0,False,2953.0,7316,3992.09
1,2026-04-01,South,P01977,General Product,Accessories,Supplier C,Warehouse 2,Pending,214.0,3419.0,2864.51,True,2.0,False,2019.0,6030,17176.62
2,2026-07-27,East,P07552,General Product,Accessories,Supplier A,Warehouse 2,Fulfilled,236.0,1576.0,6426.89,True,14.0,False,1026.0,7603,21255.0
3,2026-07-02,North,P05670,General Product,Accessories,Supplier D,Warehouse 1,Fulfilled,393.0,2021.0,6169.45,True,26.0,False,944.0,8116,9403.68
4,2025-11-02,north,P08858,Smartphone,Electronics,Supplier D,Warehouse 3,Fulfilled,732.0,1381.0,5330.58,True,1.0,False,2301.0,6320,34653.57


In [5]:
df.columns

Index(['Date', 'Region', 'Product ID', 'Product Name', 'Category', 'Supplier',
       'Warehouse', 'Order Status', 'Units Sold', 'Inventory Level',
       'Transportation Cost', 'Order Accuracy', 'Lead Time (Days)',
       'Backorder', 'Average Inventory', 'Warehouse Capacity', 'COGS'],
      dtype='str')

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10069 entries, 0 to 10068
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date                 10069 non-null  str    
 1   Region               9996 non-null   str    
 2   Product ID           10069 non-null  str    
 3   Product Name         10069 non-null  str    
 4   Category             10069 non-null  str    
 5   Supplier             10009 non-null  str    
 6   Warehouse            10009 non-null  str    
 7   Order Status         10069 non-null  str    
 8   Units Sold           10069 non-null  float64
 9   Inventory Level      10002 non-null  float64
 10  Transportation Cost  10014 non-null  str    
 11  Order Accuracy       10009 non-null  object 
 12  Lead Time (Days)     10014 non-null  float64
 13  Backorder            10069 non-null  bool   
 14  Average Inventory    10003 non-null  float64
 15  Warehouse Capacity   10069 non-null  int64  
 1

In [7]:
df.isnull().sum()

Date                    0
Region                 73
Product ID              0
Product Name            0
Category                0
Supplier               60
Warehouse              60
Order Status            0
Units Sold              0
Inventory Level        67
Transportation Cost    55
Order Accuracy         60
Lead Time (Days)       55
Backorder               0
Average Inventory      66
Warehouse Capacity      0
COGS                   56
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(52)

In [9]:
df = df.drop_duplicates()

In [10]:
df.shape

(10017, 17)

In [11]:
# Standardize text columns

df['Region'] = df['Region'].str.strip().str.title()
df['Supplier'] = df['Supplier'].str.strip().str.title()
df['Warehouse'] = df['Warehouse'].str.strip().str.title()
df['Order Status'] = df['Order Status'].str.strip().str.title()
df['Category'] = df['Category'].str.strip().str.title()

In [12]:
print("Region:")
print(df['Region'].unique())

print("\nSupplier:")
print(df['Supplier'].unique())

print("\nWarehouse:")
print(df['Warehouse'].unique())

print("\nOrder Status:")
print(df['Order Status'].unique())

print("\nCategory:")
print(df['Category'].unique())

Region:
<StringArray>
['West', 'South', 'East', 'North', nan, 'Noth', 'Sout']
Length: 7, dtype: str

Supplier:
<StringArray>
['Supplier A', 'Supplier C', 'Supplier D', 'Supplier B', nan, 'Suppplier A']
Length: 6, dtype: str

Warehouse:
<StringArray>
['Warehouse 2', 'Warehouse 1', 'Warehouse 3', nan, 'Warehoue 2']
Length: 5, dtype: str

Order Status:
<StringArray>
['Fulfilled', 'Pending', 'Canceled', 'In Transit', 'Unknown']
Length: 5, dtype: str

Category:
<StringArray>
['Electronics', 'Accessories', 'Furniture', 'Clothing']
Length: 4, dtype: str


In [13]:
# Fix spelling mistakes in categorical columns

df['Region'] = df['Region'].replace({
    'Noth': 'North',
    'Sout': 'South'
})

df['Supplier'] = df['Supplier'].replace({
    'Suppplier A': 'Supplier A'
})

df['Warehouse'] = df['Warehouse'].replace({
    'Warehoue 2': 'Warehouse 2'
})

In [14]:
print("Region:", df['Region'].unique())
print("Supplier:", df['Supplier'].unique())
print("Warehouse:", df['Warehouse'].unique())

Region: <StringArray>
['West', 'South', 'East', 'North', nan]
Length: 5, dtype: str
Supplier: <StringArray>
['Supplier A', 'Supplier C', 'Supplier D', 'Supplier B', nan]
Length: 5, dtype: str
Warehouse: <StringArray>
['Warehouse 2', 'Warehouse 1', 'Warehouse 3', nan]
Length: 4, dtype: str


In [15]:
df['Order Accuracy'] = df['Order Accuracy'].astype('boolean')

In [16]:
df['Order Accuracy'].dtype

BooleanDtype

In [17]:
# Convert numeric columns to numeric format

df['Transportation Cost'] = pd.to_numeric(
    df['Transportation Cost'],
    errors='coerce'
)
df['COGS'] = pd.to_numeric(df['COGS'], errors='coerce')

In [18]:
df[['Transportation Cost', 'COGS']].dtypes

Transportation Cost    float64
COGS                   float64
dtype: object

In [19]:
df.info()

<class 'pandas.DataFrame'>
Index: 10017 entries, 0 to 10068
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Date                 10017 non-null  str    
 1   Region               9947 non-null   str    
 2   Product ID           10017 non-null  str    
 3   Product Name         10017 non-null  str    
 4   Category             10017 non-null  str    
 5   Supplier             9957 non-null   str    
 6   Warehouse            9957 non-null   str    
 7   Order Status         10017 non-null  str    
 8   Units Sold           10017 non-null  float64
 9   Inventory Level      9951 non-null   float64
 10  Transportation Cost  9950 non-null   float64
 11  Order Accuracy       9957 non-null   boolean
 12  Lead Time (Days)     9962 non-null   float64
 13  Backorder            10017 non-null  bool   
 14  Average Inventory    9952 non-null   float64
 15  Warehouse Capacity   10017 non-null  int64  
 16  CO

In [20]:
df['Date'].isnull().sum()

np.int64(0)

In [21]:
df['Date'] = pd.to_datetime(df['Date'], errors='coerce')

In [22]:
df['Date'].dtype

dtype('<M8[us]')

In [23]:
df['Date'].isnull().sum()

np.int64(180)

In [24]:
# Reload original Date column
original_dates = pd.read_csv("Inventory_SupplyChain.csv")['Date'].astype('string').str.strip()

# Create empty datetime column
parsed_dates = pd.Series(pd.NaT, index=original_dates.index, dtype='datetime64[ns]')

# 1. YYYY-MM-DD
mask = original_dates.str.match(r'^\d{4}-\d{1,2}-\d{1,2}$', na=False)
parsed_dates.loc[mask] = pd.to_datetime(
    original_dates.loc[mask],
    format='%Y-%m-%d'
)

# 2. YYYY/MM/DD
mask = original_dates.str.match(r'^\d{4}/\d{1,2}/\d{1,2}$', na=False)
parsed_dates.loc[mask] = pd.to_datetime(
    original_dates.loc[mask],
    format='%Y/%m/%d'
)

# 3. DD/MM/YYYY
mask = original_dates.str.match(r'^\d{1,2}/\d{1,2}/\d{4}$', na=False)
parsed_dates.loc[mask] = pd.to_datetime(
    original_dates.loc[mask],
    format='%d/%m/%Y'
)

# 4. MM-DD-YYYY
mask = original_dates.str.match(r'^\d{1,2}-\d{1,2}-\d{4}$', na=False)
parsed_dates.loc[mask] = pd.to_datetime(
    original_dates.loc[mask],
    format='%m-%d-%Y'
)

# Put correctly parsed dates back into df
df['Date'] = parsed_dates.loc[df.index]

In [25]:
df['Date'].isnull().sum()

np.int64(0)

In [26]:
df['Date'].dtype

dtype('<M8[ns]')

In [27]:
print("Minimum Date:", df['Date'].min())
print("Maximum Date:", df['Date'].max())

Minimum Date: 2025-01-01 00:00:00
Maximum Date: 2026-08-31 00:00:00


In [28]:
df.isnull().sum()

Date                    0
Region                 70
Product ID              0
Product Name            0
Category                0
Supplier               60
Warehouse              60
Order Status            0
Units Sold              0
Inventory Level        66
Transportation Cost    67
Order Accuracy         60
Lead Time (Days)       55
Backorder               0
Average Inventory      65
Warehouse Capacity      0
COGS                   67
dtype: int64

In [29]:
# Missing value percentage
missing_percent = (df.isnull().sum() / len(df)) * 100

print(missing_percent.round(2))

Date                   0.00
Region                 0.70
Product ID             0.00
Product Name           0.00
Category               0.00
Supplier               0.60
Warehouse              0.60
Order Status           0.00
Units Sold             0.00
Inventory Level        0.66
Transportation Cost    0.67
Order Accuracy         0.60
Lead Time (Days)       0.55
Backorder              0.00
Average Inventory      0.65
Warehouse Capacity     0.00
COGS                   0.67
dtype: float64


In [30]:
missing_summary = pd.DataFrame({
    'Missing Values': df.isnull().sum(),
    'Missing %': (df.isnull().sum() / len(df) * 100).round(2)
})

missing_summary = missing_summary[missing_summary['Missing Values'] > 0]

missing_summary

,Missing Values,Missing %
Region,70,0.70
Supplier,60,0.60
Warehouse,60,0.60
Inventory Level,66,0.66
Transportation Cost,67,0.67
Order Accuracy,60,0.60
Lead Time (Days),55,0.55
Average Inventory,65,0.65
COGS,67,0.67


In [31]:
# Fill missing categorical values with mode

for col in ['Region', 'Supplier', 'Warehouse']:
    df[col] = df[col].fillna(df[col].mode()[0])

In [32]:
df[['Region', 'Supplier', 'Warehouse']].isnull().sum()

Region       0
Supplier     0
Warehouse    0
dtype: int64

In [33]:
# Fill missing numerical values with median

numeric_cols = [
    'Inventory Level',
    'Transportation Cost',
    'Lead Time (Days)',
    'Average Inventory',
    'COGS'
]

for col in numeric_cols:
    df[col] = df[col].fillna(df[col].median())

In [34]:
df[numeric_cols].isnull().sum()

Inventory Level        0
Transportation Cost    0
Lead Time (Days)       0
Average Inventory      0
COGS                   0
dtype: int64

In [35]:
# Fill missing Order Accuracy with mode

df['Order Accuracy'] = df['Order Accuracy'].fillna(
    df['Order Accuracy'].mode()[0]
)

In [36]:
df['Order Accuracy'].isnull().sum()

np.int64(0)

In [37]:
df['Order Accuracy'].value_counts()

Order Accuracy
True     9178
False     839
Name: count, dtype: Int64

In [38]:
# Final missing value check
df.isnull().sum()

Date                   0
Region                 0
Product ID             0
Product Name           0
Category               0
Supplier               0
Warehouse              0
Order Status           0
Units Sold             0
Inventory Level        0
Transportation Cost    0
Order Accuracy         0
Lead Time (Days)       0
Backorder              0
Average Inventory      0
Warehouse Capacity     0
COGS                   0
dtype: int64

In [39]:
# Check dataset shape
df.shape

(10017, 17)

In [40]:
# Check data types
df.dtypes

Date                   datetime64[ns]
Region                            str
Product ID                        str
Product Name                      str
Category                          str
Supplier                          str
Warehouse                         str
Order Status                      str
Units Sold                    float64
Inventory Level               float64
Transportation Cost           float64
Order Accuracy                boolean
Lead Time (Days)              float64
Backorder                        bool
Average Inventory             float64
Warehouse Capacity              int64
COGS                          float64
dtype: object

In [41]:
# Statistical summary of numerical columns
df.describe()

,Date,Units Sold,Inventory Level,Transportation Cost,Lead Time (Days),Average Inventory,Warehouse Capacity,COGS
count,10017,10017.000000,10017.000000,10017.000000,10017.000000,10017.000000,10017.000000,10017.000000
mean,2025-10-31 11:15:30.458221056,530.970350,2600.551463,7935.577773,16.206150,2517.799491,7407.418289,37462.474812
min,2025-01-01 00:00:00,40.000000,-10.000000,-50.000000,0.000000,202.000000,5002.000000,-100.000000
25%,2025-06-01 00:00:00,288.000000,1274.000000,4149.420000,8.000000,1783.000000,6107.000000,16477.200000
50%,2025-10-29 00:00:00,511.000000,2612.000000,7873.460000,16.000000,2473.500000,7420.000000,30752.995000
75%,2026-04-03 00:00:00,757.000000,3750.000000,11491.920000,23.000000,3246.000000,8624.000000,52704.450000
max,2026-08-31 00:00:00,1241.000000,8000.000000,17940.070000,75.000000,5778.000000,9995.000000,146599.790000
std,NaN,288.488857,1518.503231,4391.590240,9.144743,1051.580660,1448.380993,26954.251721


In [42]:
# Check negative values in important numeric columns

print("Negative Inventory Level:", (df['Inventory Level'] < 0).sum())
print("Negative Transportation Cost:", (df['Transportation Cost'] < 0).sum())
print("Negative COGS:", (df['COGS'] < 0).sum())

Negative Inventory Level: 14
Negative Transportation Cost: 13
Negative COGS: 12


In [43]:
# Show rows with negative values

df[
    (df['Inventory Level'] < 0) |
    (df['Transportation Cost'] < 0) |
    (df['COGS'] < 0)
]

,Date,Region,Product ID,Product Name,Category,Supplier,Warehouse,Order Status,Units Sold,Inventory Level,Transportation Cost,Order Accuracy,Lead Time (Days),Backorder,Average Inventory,Warehouse Capacity,COGS
520,2025-03-05,East,P01354,Tablet,Electronics,Supplier B,Warehouse 1,Fulfilled,489.0,401.0,-25.00,True,15.0,True,609.0,6755,37699.64
605,2025-06-23,East,P06736,Mouse,Electronics,Supplier C,Warehouse 3,Canceled,386.0,-10.0,10126.63,False,33.0,False,2414.0,9765,11733.57
746,2026-06-30,North,P00821,General Product,Accessories,Supplier A,Warehouse 1,Fulfilled,697.0,-10.0,12195.70,True,32.0,False,2340.0,5027,53675.35
968,2026-01-31,East,P00784,Conference Table,Furniture,Supplier C,Warehouse 2,Fulfilled,605.0,-10.0,3255.50,True,18.0,False,5301.0,5109,26158.26
1532,2026-02-06,South,P01081,Mouse,Electronics,Supplier B,Warehouse 3,Fulfilled,72.0,384.0,-25.00,True,8.0,False,1256.0,8898,3627.31
1595,2026-03-07,East,P04060,Smartphone,Electronics,Supplier A,Warehouse 3,Fulfilled,133.0,-5.0,1969.54,True,10.0,False,2181.0,5824,6685.19
1784,2026-01-16,South,P02052,Bookshelf,Furniture,Supplier A,Warehouse 1,Pending,96.0,2960.0,-50.00,True,23.0,False,1929.0,9665,3753.73
1866,2026-08-07,North,P04748,Monitor,Electronics,Supplier C,Warehouse 2,Fulfilled,88.0,3973.0,-50.00,True,15.0,True,2393.0,5486,2181.55
1878,2025-02-18,North,P08238,General Product,Accessories,Supplier A,Warehouse 1,Fulfilled,221.0,3302.0,4369.47,True,17.0,False,3236.0,7363,-50.00
1899,2026-07-28,North,P03834,Mouse,Electronics,Supplier C,Warehouse 1,Fulfilled,470.0,2579.0,4562.12,False,22.0,False,1593.0,8072,-100.00


In [44]:
# Count rows having one or more negative values

negative_rows = (
    (df['Inventory Level'] < 0) |
    (df['Transportation Cost'] < 0) |
    (df['COGS'] < 0)
)

print("Rows with negative values:", negative_rows.sum())

Rows with negative values: 39


In [45]:
# Check how many negative issues each row has

(
    (df[['Inventory Level', 'Transportation Cost', 'COGS']] < 0)
    .sum(axis=1)
    .value_counts()
    .sort_index()
)

0    9978
1      39
Name: count, dtype: int64

In [46]:
# Check which column contains the negative value

negative_data = df.loc[
    negative_rows,
    ['Inventory Level', 'Transportation Cost', 'COGS']
]

negative_data

,Inventory Level,Transportation Cost,COGS
520,401.0,-25.00,37699.64
605,-10.0,10126.63,11733.57
746,-10.0,12195.70,53675.35
968,-10.0,3255.50,26158.26
1532,384.0,-25.00,3627.31
1595,-5.0,1969.54,6685.19
1784,2960.0,-50.00,3753.73
1866,3973.0,-50.00,2181.55
1878,3302.0,4369.47,-50.00
1899,2579.0,4562.12,-100.00


In [47]:
# Remove rows with invalid negative values

df = df[
    (df['Inventory Level'] >= 0) &
    (df['Transportation Cost'] >= 0) &
    (df['COGS'] >= 0)
]

In [48]:
df.shape

(9978, 17)

In [49]:
print("Negative Inventory Level:", (df['Inventory Level'] < 0).sum())
print("Negative Transportation Cost:", (df['Transportation Cost'] < 0).sum())
print("Negative COGS:", (df['COGS'] < 0).sum())

Negative Inventory Level: 0
Negative Transportation Cost: 0
Negative COGS: 0


In [50]:
df.duplicated().sum()

np.int64(25)

In [52]:
df = df.drop_duplicates()

In [53]:
df.duplicated().sum()

np.int64(0)

In [54]:
df.shape

(9953, 17)

In [55]:
df.isnull().sum()

Date                   0
Region                 0
Product ID             0
Product Name           0
Category               0
Supplier               0
Warehouse              0
Order Status           0
Units Sold             0
Inventory Level        0
Transportation Cost    0
Order Accuracy         0
Lead Time (Days)       0
Backorder              0
Average Inventory      0
Warehouse Capacity     0
COGS                   0
dtype: int64

# Data Cleaning Summary

The Inventory & Supply Chain dataset was cleaned and prepared for further analysis and visualization.

### 1. Initial Dataset

* The dataset initially contained **10,069 rows and 17 columns**.
* The dataset includes information related to dates, regions, products, suppliers, warehouses, order status, inventory, transportation cost, lead time, backorders, warehouse capacity, and COGS.

### 2. Duplicate Records

* **52 duplicate rows** were identified initially and removed.
* After further data cleaning, **25 additional duplicate rows** were identified and removed.
* The final duplicate check confirmed **0 duplicate rows**.

### 3. Text Standardization

Text-based categorical columns were standardized using trimming and title-case formatting. The following columns were cleaned:

* Region
* Supplier
* Warehouse
* Order Status
* Category

Spelling inconsistencies such as `Noth`, `Sout`, `Suppplier A`, and `Warehoue 2` were corrected to their proper values.

### 4. Missing Values

Missing values were identified across several columns.

* Categorical missing values in **Region, Supplier, and Warehouse** were filled using the **mode**.
* Numerical missing values in **Inventory Level, Transportation Cost, Lead Time, Average Inventory, and COGS** were filled using the **median**.
* Missing values in **Order Accuracy** were filled using its **mode**.
* The final missing-value check confirmed **0 missing values across all 17 columns**.

### 5. Data Types

Data types were corrected according to the nature of each variable:

* `Date` → datetime
* `Order Accuracy` → boolean
* `Backorder` → boolean
* Transportation Cost and COGS → numeric
* Inventory, lead time, average inventory and other quantitative fields → numeric

The final dataset contains appropriate data types for analysis.

### 6. Date Cleaning

The Date column contained multiple date formats. Different formats including:

* `YYYY-MM-DD`
* `YYYY/MM/DD`
* `DD/MM/YYYY`
* `MM-DD-YYYY`

were handled and converted into a consistent datetime format.

The final date range is from **January 1, 2025 to August 31, 2026**, with no invalid or missing dates remaining.

### 7. Invalid Negative Values

Negative values were found in columns where negative values are not logically valid:

* Inventory Level
* Transportation Cost
* COGS

A total of **39 rows** contained one or more negative values. These invalid records were removed rather than replacing them with artificial values. After removal, all three columns contained **zero negative values**.

### 8. Final Dataset

After completing the cleaning process:

* **Final rows:** 9,953
* **Columns:** 17
* **Duplicate rows:** 0
* **Missing values:** 0
* **Negative invalid values:** 0

The dataset is now clean, consistent, and ready for **Exploratory Data Analysis (EDA), KPI analysis, visualization, and Power BI dashboard development**.



In [56]:
df.to_csv("Inventory_SupplyChain_Cleaned.csv", index=False)